File: src/backupsubscription.ts
import { AppContext } from "./appcontext";
import { UnsupportedOperationException } from "./exception/illegalargumentexception";
import { CompletionException } from "./exception/unsupportedoperationexception";
import { VaultInfoResponseBody } from "./network/response/vaultinforesponsebody";
import { Order } from "./payment/order";
import { PricingPlan } from "./payment/pricingplan";
import { Receipt } from "./payment/receipt";
import { PaymentService } from "./service/paymentservice";
import { SubscriptionService } from "./service/subscriptionservice";
import { ServiceEndpoint } from "./serviceendpoint";
import { HttpExceptionHandler } from "./vault/httpexceptionhandler";
import { PaymentServiceRender } from "./vault/paymentservicerender";
import { SubscriptionServiceRender } from "./vault/subscriptionservicerender";

export class BackupSubscription extends ServiceEndpoint implements SubscriptionService<BackupSubscription.BackupInfo>, PaymentService {
	private subscriptionService: SubscriptionServiceRender;
	private paymentService: PaymentServiceRender;

	public constructor(context: AppContext, providerAddress: string) {
		super(context, providerAddress);
		this.paymentService = new PaymentServiceRender(this);
		this.subscriptionService = new SubscriptionServiceRender(this);
	}

	public subscribe(pricingPlan: string): Promise<BackupSubscription.BackupInfo> {
		return this.promiseWithConvertedException<BackupSubscription.BackupInfo>(async ()=>{
			await this.subscriptionService.subscribeBackup();
			return this.getBackupInfoByResponseBody(this.subscriptionService.getBackupVaultInfo());
		});
	}

	private getBackupInfoByResponseBody(body: VaultInfoResponseBody): BackupSubscription.BackupInfo {
		return new BackupSubscription.BackupInfo().setDid(body.getDid())
				.setMaxStorage(body.getMaxStorage())
				.setFileUseStorage(body.getFileUseStorage())
				.setDbUseStorage(body.getDbUseStorage())
				.setModifyTime(body.getModifyTimeStr())
				.setStartTime(body.getStartTimeStr())
				.setEndTime(body.getEndTimeStr())
				.setPricingUsing(body.getPricingUsing())
				.setIsExisting(body.isExisting());
	}

	public unsubscribe(): Promise<void> {
		throw new UnsupportedOperationException();
	}

	public activate(): Promise<void> {
		throw new UnsupportedOperationException();
	}

	public deactivate(): Promise<void> {
		throw new UnsupportedOperationException();
	}

	public checkSubscription(): Promise<BackupSubscription.BackupInfo> {
		return this.promiseWithConvertedException<BackupSubscription.BackupInfo>(async ()=>{
			return this.getBackupInfoByResponseBody(this.subscriptionService.getBackupVaultInfo());
		});
	}

	public getPricingPlanList(): Promise<PricingPlan[]> {
		return this.promiseWithConvertedException<PricingPlan[]>(()=>{
			return this.paymentService.getBackupPlanList();
		});
	}

	public getPricingPlan(planName: string): Promise<PricingPlan> {
		return this.promiseWithConvertedException<PricingPlan>(()=>{
			return this.paymentService.getBackupPlan(planName);
		});
	}

	public placeOrder(planName: string): Promise<Order> {
		return this.promiseWithConvertedException<Order>(()=>{
			return this.paymentService.getOrderInfo(this.paymentService.createBackupOrder(planName));
		});
	}

	public getOrder(orderId: string): Promise<Order> {
		return this.promiseWithConvertedException<Order>(()=>{
			return this.paymentService.getOrderInfo(orderId);
		});
	}

	public payOrder(orderId: string, transIds: string[]): Promise<Receipt> {
		return this.promiseWithConvertedException<Receipt>(()=>{
			this.paymentService.payOrder(orderId, transIds);
			//TODO:
			return null;
		});
	}

	public getReceipt(receiptId: string): Promise<Receipt> {
		throw new UnsupportedOperationException();
	}
}

export namespace BackupSubscription {
	export class BackupInfo {
		private did: string;
		private maxStorage: number;
		private fileUseStorage: number;
		private dbUseStorage: number;
		private modifyTime: string;
		private startTime: string;
		private endTime: string;
		private pricingUsing: string;
		private isExisting: boolean;

		public getDid(): string {
			return this.did;
		}

		public setDid(did: string): BackupInfo {
			this.did = did;
			return this;
		}

		public getMaxStorage(): number {
			return this.maxStorage;
		}

		public setMaxStorage(maxStorage: number): BackupInfo {
			this.maxStorage = maxStorage;
			return this;
		}

		public getFileUseStorage(): number {
			return this.fileUseStorage;
		}

		public setFileUseStorage(fileUseStorage: number): BackupInfo {
			this.fileUseStorage = fileUseStorage;
			return this;
		}

		public getDbUseStorage(): number {
			return this.dbUseStorage;
		}

		public setDbUseStorage(dbUseStorage: number): BackupInfo {
			this.dbUseStorage = dbUseStorage;
			return this;
		}

		public getModifyTime(): string {
			return this.modifyTime;
		}

		public setModifyTime(modifyTime: string): BackupInfo {
			this.modifyTime = modifyTime;
			return this;
		}

		public getStartTime(): string {
			return this.startTime;
		}

		public setStartTime(startTime: string): BackupInfo {
			this.startTime = startTime;
			return this;
		}

		public getEndTime(): string {
			return this.endTime;
		}

		public setEndTime(endTime: string): BackupInfo {
			this.endTime = endTime;
			return this;
		}

		public getPricingUsing(): string {
			return this.pricingUsing;
		}

		public setPricingUsing(pricingUsing: string): BackupInfo {
			this.pricingUsing = pricingUsing;
			return this;
		}

		public getIsExisting(): boolean {
			return this.isExisting;
		}

		public setIsExisting(isExisting: boolean): BackupInfo {
			this.isExisting = isExisting;
			return this;
		}
	}
}

File: src/java/org/elastos/hive/BackupSubscription.java
package org.elastos.hive;

import org.elastos.hive.exception.HiveException;
import org.elastos.hive.exception.NotImplementedException;
import org.elastos.hive.subscription.BackupInfo;
import org.elastos.hive.subscription.PricingPlan;
import org.elastos.hive.subscription.SubscriptionController;
import org.elastos.hive.subscription.payment.Order;
import org.elastos.hive.subscription.payment.PaymentController;
import org.elastos.hive.subscription.payment.Receipt;

import org.elastos.hive.service.PaymentService;
import org.elastos.hive.service.SubscriptionService;

import java.util.List;
import java.util.concurrent.CompletableFuture;
import java.util.concurrent.CompletionException;

public class BackupSubscription extends ServiceContext
				implements SubscriptionService<BackupInfo>, PaymentService {

	private SubscriptionController subscriptionController;
	private PaymentController paymentController;

	public BackupSubscription(AppContext context, String providerAddress) throws HiveException {
		super(context, providerAddress);
		subscriptionController = new SubscriptionController(this);
		paymentController = new PaymentController(this);
	}

	@Override
	public CompletableFuture<List<PricingPlan>> getPricingPlanList() {
		return CompletableFuture.supplyAsync(()-> {
			try {
				return subscriptionController.getBackupPricingPlanList();
			} catch (RuntimeException | HiveException e) {
				throw new CompletionException(e);
			}
		});
	}

	@Override
	public CompletableFuture<PricingPlan> getPricingPlan(String planName) {
		return CompletableFuture.supplyAsync(()-> {
			if (planName == null)
				throw new IllegalArgumentException("Empty plan name");

			try {
				return subscriptionController.getBackupPricingPlan(planName);
			} catch (RuntimeException | HiveException e) {
				throw new CompletionException(e);
			}
		});
	}

	public CompletableFuture<BackupInfo> subscribe() {
		return this.subscribe(null);
	}

	@Override
	public CompletableFuture<BackupInfo> subscribe(String credential) {
		return CompletableFuture.supplyAsync(()-> {
			if (credential != null)
				throw new NotImplementedException("Paid pricing plan will be supported later");

			try {
				return subscriptionController.subscribeToBackup(null);
			} catch (HiveException | RuntimeException e) {
				throw new CompletionException(e);
			}
		});
	}

	@Override
	public CompletableFuture<Void> unsubscribe() {
		return CompletableFuture.runAsync(()-> {
			try {
				subscriptionController.unsubscribeBackup();
			} catch (HiveException | RuntimeException e) {
				throw new CompletionException(e);
			}
		});
	}

	@Override
	public CompletableFuture<BackupInfo> checkSubscription() {
		return CompletableFuture.supplyAsync(()-> {
			try {
				return subscriptionController.getBackupInfo();
			} catch (HiveException | RuntimeException e) {
				throw new CompletionException(e);
			}
		});
	}

	@Override
	public CompletableFuture<Order> placeOrder(String planName) {
		return CompletableFuture.supplyAsync(() -> {
			try {
				return paymentController.placeOrder("backup", planName);
			} catch (HiveException | RuntimeException e) {
				throw new CompletionException(e);
			}
		});
	}

	@Override
	public CompletableFuture<Order> getOrder(String orderId) {
		return CompletableFuture.supplyAsync(() -> {
			try {
				return paymentController.getOrder(orderId);
			} catch (HiveException | RuntimeException e) {
				throw new CompletionException(e);
			}
		});
	}

	@Override
	public CompletableFuture<Receipt> payOrder(String orderId, String transactionId) {
		return CompletableFuture.supplyAsync(() -> {
			try {
				return paymentController.payOrder(orderId, transactionId);
			} catch (HiveException | RuntimeException e) {
				throw new CompletionException(e);
			}
		});
	}

	@Override
	public CompletableFuture<List<Order>> getOrderList() {
		return CompletableFuture.supplyAsync(() -> {
			try {
				return paymentController.getOrders("backup");
			} catch (HiveException | RuntimeException e) {
				throw new CompletionException(e);
			}
		});
	}

	@Override
	public CompletableFuture<Receipt> getReceipt(String orderId) {
		return CompletableFuture.supplyAsync(() -> {
			try {
				return paymentController.getReceipt(orderId);
			} catch (HiveException | RuntimeException e) {
				throw new CompletionException(e);
			}
		});
	}

	@Override
	public CompletableFuture<String> getVersion() {
		return CompletableFuture.supplyAsync(() -> {
			try {
				return paymentController.getVersion();
			} catch (HiveException | RuntimeException e) {
				throw new CompletionException(e);
			}
		});
	}
}
